casper/src/main/rholang/MakeMint.rho

```diff
@@ -47,18 +49,27 @@ in {
                 }
               }
             } |
-            contract thisPurse(@"deposit", @amount, @src, success) = {
-              new result in {
-                @src!(*decr, amount, *result) | //call src decr function.
-                for(@decrSuccess <- result) {
-                  if (decrSuccess) {
-                    balance!("add", amount, *success) // add transferred amount to this purse
+
+            contract thisPurse(@"deposit", @amount, @src, successCh) = {
+              new ch in {
+                balance!("add", amount, *ch) |
+                for(@addSuccess <- ch) {
+                  if(addSuccess) {
+                    @src!(*decr, amount, *ch) |
```

Comment:
All possible failure modes have to be checked **before** the call to "decr".

See https://github.com/Agoric/PlaygroundVat/blob/master/examples/contract/makeMint.js#L42

Reply:
what other failure modes are you thinking about?

Reply:
Not other failure modes but the order of checking.

I'm saying something similar to what Kent is saying: check before changing any balances. Don't change and then undo.

Reply:
Also... we need to make sure there is no observable state in between increasing the destination balance and decreasing the source balance. Somehow we need to do something like:

```
  for (@destBal <- destBalCh; @srcBal <- srcBalCh) {
    // check that srcBal >= amount and that destBal + amount doesn't overflow
    if (okToCommit) {
      destBalCh!(destBal + amount) | srcBalCh!(srcBal - amount)
    } else {
      destBal!(destBal) | srcBal!(srcBal)
    }
  }
```

In fact, when we get peek, we shouldn't even consume the balances from their channels until the commit point.

Reply:
>> we need to make sure there is no observable state in between 
This is a very good point!

Reply:
probably should be its own JIRA ticket

Reply:
Still processing...

Reply:
I've created a new JIRA ticket for this: https://rchain.atlassian.net/browse/RCHAIN-2926